Sets bCapture* true only while showing subwindow

nznaza opened the pull request:
Sets bAlwaysPersistRenderingState to true. This fixes bugs related to capture while on NoDisplay.
Sets bCaptureOnMovement and bCaptureEveryFrame to true only when they are going to be viewed on a subwindow

<!-- Thank you for submitting a pull request! -->
<!-- ⚠️⚠️ Do Not Delete This! pull_request_template ⚠️⚠️ -->
<!-- Please read our contribution guidelines: https://microsoft.github.io/AirSim/CONTRIBUTING/ -->

Fixes: #1724 
Fixes: #2967
Fixes: #3504
Fixes: #3070

Partially fixes #4112

Should fix most of the FPS while capturing problems.

## About
<!-- Describe what your PR is about. -->
This PR fixes the NoDisplay/Headless bug by setting bAlwaysPersistRenderingState to true.

Setting bCaptureOnMovement and bCaptureEveryFrame to true only while it's on an active subwindow camera improves the performance of image capture/recording and python simGetImages

## How Has This Been Tested?
<!-- Please, describe how you have tested your changes to help us incorporate them. -->
A python script was used to display the captured images to corroborate the accurate capture of the desired captures; the same script was used for all captures using ComputerVision.

```
import airsim
from cv2 import data
import numpy as np
from airsim import *
import cv2 
import datetime

def ImageResponsetoImage(selected):
    isFloat = selected.pixels_as_float
    if isFloat:
        img1d = np.array(selected.image_data_float)
    else:
        img1d = np.frombuffer(selected.image_data_uint8, dtype=np.uint8)

    width = selected.width
    height = selected.height
    channels = int(img1d.size / (height * width))
    img = img1d.reshape(height, width, channels)
    if (isFloat):
        depthinmm = img*1000
        depth_16bit =  np.clip(depthinmm, 0,65535) 
        return depth_16bit.astype(np.uint16)

    else:
        return img.astype(np.uint8)

if __name__ == "__main__":
    client = MultirotorClient()
    client.confirmConnection()

    i = 0
    inc = True
    while True:
        responses = client.simGetImages(
            [
                ImageRequest("front_left", ImageType.Scene, False, False),
                ImageRequest("front_left", ImageType.DepthPerspective, True, False),
                ImageRequest("front_left", ImageType.Segmentation, False, False),
                
                ImageRequest("front_center", ImageType.Scene, False, False),
                ImageRequest("front_center", ImageType.DepthPerspective, True, False),
                ImageRequest("front_center", ImageType.Segmentation, False, False),
                
                ImageRequest("front_right", ImageType.Scene, False, False),
                ImageRequest("front_right", ImageType.DepthPerspective, True, False),
                ImageRequest("front_right", ImageType.Segmentation, False, False),
                
                ImageRequest("bottom_center", ImageType.Scene, False, False),
                ImageRequest("bottom_center", ImageType.DepthPerspective, True, False),
                ImageRequest("bottom_center", ImageType.Segmentation, False, False),

                ImageRequest("back_center", ImageType.Scene, False, False),
                ImageRequest("back_center", ImageType.DepthPerspective, True, False),
                ImageRequest("back_center", ImageType.Segmentation, False, False),
                
            ]
        )
        tm = datetime.datetime.now().second % 10
        
        if tm % 2 == 0 :
            if inc:
                i += 1
            inc = False
        else:
            inc = True   
        if i >= len(responses):
            i = 0
        selected = responses[i]

        cv2.imshow("", ImageResponsetoImage(selected))
        cv2.waitKey(1)

```

Performance goes back up after stopping capture while on NoDisplay, but stays the same after recording or using the python simGetImages, but this can be fixed by displaying and hiding a subwindow.

settings.json

```
{
  "SeeDocsAt": "https://github.com/Microsoft/AirSim/blob/master/docs/settings.md",
  "SettingsVersion": 1.2,
  "ViewMode": "NoDisplay",
  "SimMode": "ComputerVision",
  "Recording": {
    "RecordOnMove": true,
    "RecordInterval": 0.05,
    "Folder": "",
    "Enabled": false,
    "Cameras": [
        { "CameraName": "0", "ImageType": 0, "PixelsAsFloat": false,  "VehicleName": "ComputerVision", "Compress": true },
        { "CameraName": "0", "ImageType": 1, "PixelsAsFloat": false,  "VehicleName": "ComputerVision", "Compress": true },
        { "CameraName": "0", "ImageType": 2, "PixelsAsFloat": false,  "VehicleName": "ComputerVision", "Compress": true },
        { "CameraName": "0", "ImageType": 7, "PixelsAsFloat": false,  "VehicleName": "ComputerVision", "Compress": true }
    ]
  },
  "CameraDefaults": {
    "CaptureSettings": [
      {
        "ImageType": 0,
        "Width": 640,
        "Height": 480
      },
      {
        "ImageType": 2,
        "Width": 640,
        "Height": 480
      },
      {
        "ImageType": 7,
        "Width": 640,
        "Height": 480
      },
      {
        "ImageType": 8,
        "Width": 640,
        "Height": 480
      },
      {
        "ImageType": 9,
        "Width": 640,
        "Height": 480
      },
      {
        "ImageType": 5,
        "Width": 640,
        "Height": 480
      }
    ],
    "NoiseSettings": [
      {
        "Enabled": false,
        "ImageType": 7,

        "RandContrib": 0.2,
        "RandSpeed": 100000.0,
        "RandSize": 500.0,
        "RandDensity": 2,

        "HorzWaveContrib":0.03,
        "HorzWaveStrength": 0.08,
        "HorzWaveVertSize": 1.0,
        "HorzWaveScreenSize": 1.0,

        "HorzNoiseLinesContrib": 1.0,
        "HorzNoiseLinesDensityY": 0.01,
        "HorzNoiseLinesDensityXY": 0.5,

        "HorzDistortionContrib": 1.0,
        "HorzDistortionStrength": 0.002
      }
    ]
  },
  "SegmentationSettings": {
    "InitMethod": "",
    "MeshNamingMethod": "StaticMeshName",
    "OverrideExisting": true
  }
}

```

## Screenshots (if appropriate)

NoDisplay

RGB
Previous:
![PreFixRGB](https://user-images.githubusercontent.com/5291065/140605085-a3ec1f09-e423-4106-8b06-b57120fa5a87.png)

![FixRGB](https://user-images.githubusercontent.com/5291065/140605082-863f01eb-d6cd-4952-a773-2037f3c02476.png)

Depth:
Previous:
![PreFixDepth](https://user-images.githubusercontent.com/5291065/140605084-7d105378-3a78-412d-ad4e-1f1020aae8a6.png)

After:
![FixDepth](https://user-images.githubusercontent.com/5291065/140605081-e8fd9fc5-6692-4f8e-8431-11a08dfa3c8c.png)

Segmentation 
Previous:
![PreFixSegmentation](https://user-images.githubusercontent.com/5291065/140605080-9ee7ad76-286a-40b0-bbdf-6eabc86986fe.png)

After:
![FixSegmentation](https://user-images.githubusercontent.com/5291065/140605083-46730413-62f2-4547-a97d-f8d47abb3c9a.png)

Performance captures

Recording previous to fix
![Recording prefix](https://user-images.githubusercontent.com/5291065/140610499-04d8fd04-3b91-4f2f-b8b5-9daebeaa3b14.png)

After recording and showing and hiding subwindow (small gain)
![Recording preFix aSubwindows](https://user-images.githubusercontent.com/5291065/140610496-6a32ffed-9048-4b8a-bb9e-78fe95f1a959.png)

Recording after fix (Lower timings, )
![Recording after fix](https://user-images.githubusercontent.com/5291065/140610653-d20fb735-f600-4ff0-a3d4-00e1d995e78d.png)

After recording and showing and hiding subwindow (Back to normal)
![Recording after Fix Subwindows](https://user-images.githubusercontent.com/5291065/140610709-613b50e0-631d-43a2-86c5-3cc440524abe.png)

Python capture before fix
![python capture before](https://user-images.githubusercontent.com/5291065/140611029-039b7157-ea96-48b5-bd1b-96c63c4ebb70.png)

Python capture post fix
![python capture](https://user-images.githubusercontent.com/5291065/140610868-50f1a8a5-c65d-4f9b-ace9-9ce68b7c23b3.png)

Python capture before fix (Only front_center)
![python before only front](https://user-images.githubusercontent.com/5291065/140611168-e03e660f-9069-429b-88bf-03009332ffe5.png)

Python capture after fix (Only front_center)
![python after only front](https://user-images.githubusercontent.com/5291065/140611166-fd5e0fe7-abf5-42ab-be18-d17c465880ca.png)

Python capture before fix (Only front_center Scene)
![only RGB](https://user-images.githubusercontent.com/5291065/140611272-f43d2dbd-78d1-42e3-b561-8278c35b0247.png)

Python capture after fix (Only front_center Scene)
![after only RGB](https://user-images.githubusercontent.com/5291065/140611227-dc3e2e62-17f4-40b5-8c3b-ce15e732d909.png)

While capturing with python's simGetImages the performance after the fix are lower, but measurable: 205 ms vs 185ms, 70ms vs 54, single images are within margin error, but fix is still lower

For comparison here are the timings of all cameras and NoDIsplay

While capturing
![capture python ND](https://user-images.githubusercontent.com/5291065/140612125-a93de350-684b-4e07-b3d7-6d56877087f1.png)

After stopping capturing, the levels go back to normal without needing to enable a subwindow.
![After capture python ND](https://user-images.githubusercontent.com/5291065/140612126-365eb920-3e59-4803-96e6-2c50804c5e71.png)

Commit:
Sets bCapture* true only while showing subwindow

Sets bAlwaysPersistRenderingState to true. This fixes bugs of capture
Sets bCaptureOnMovement and bCaptureEveryFrame to true only when they are going to be viewed on as subwindow

rajat2004 commented on Unreal/Plugins/AirSim/Source/CameraDirector.cpp:
```diff
@@ -344,21 +344,16 @@ void ACameraDirector::notifyViewModeChanged()
-    else {
-        gameViewport->bDisableWorldRendering = 0;
-    }
+    gameViewport->bDisableWorldRendering = (uint32)mode_;
```
You can just use the `nodisplay` variable instead of casting `mode_`

nznaza replied:
Correct, I was removing the branch assignation and used the wrong variable, will issue a commit to address this, thanks.

rajat2004 commented on Unreal/Plugins/AirSim/Source/PIPCamera.cpp:
```diff
@@ -98,7 +98,7 @@ void APIPCamera::BeginPlay()
     }
 
-    onViewModeChanged(false);
+    onViewModeChanged(true);
```
So in the beginning, we set all cameras to NoDisplay mode to improve performance?
 Would be great if you could add some comments in the code also as to why this helps

nznaza replied:
We initialize all cameras with bCaptureOnMovement and bCaptureEveryFrame to false.

This improves performance because the capture components are no longer updating every frame and only update while requesting an image (and are set to true while we are showing them on a sub-window because sub-window captures don’t count as a request)

rajat2004 commented on Unreal/Plugins/AirSim/Source/CameraDirector.cpp:
```diff
@@ -344,21 +344,16 @@ void ACameraDirector::notifyViewModeChanged()
-        ExternalCamera->onViewModeChanged(nodisplay);
-    if (front_camera_)
-        front_camera_->onViewModeChanged(nodisplay);
+    //if (fpv_camera_)
+    //    fpv_camera_->onViewModeChanged(nodisplay);
+    //if (backup_camera_)
+    //    backup_camera_->onViewModeChanged(nodisplay);
+    //if (ExternalCamera)
+    //    ExternalCamera->onViewModeChanged(nodisplay);
+    //if (front_camera_)
+    //    front_camera_->onViewModeChanged(nodisplay);
```
Why are these no longer needed? And it's better to just remove the code rather then keep commented out pieces

nznaza replied:
Because cameras all cameras start with bCaptureOnMovement and bCaptureEveryFrame set to false regardless of nodisplay on BeginPlay(), and only required to be set true if there are going to be shown on a sub-window

I didn’t remove the code from commit while testing for regressions, but this is indeed no longer needed

rajat2004 reviewed (commented):
Still need to test, but any idea why the performance for simGetImages dropped?

nznaza commented:
> Still need to test, but any idea why the performance for simGetImages dropped?

Sorry I don’t quite get the question, if you mean why is performance still drops while simGetImages is called: it’s because we still need to acquire the images (this is what causes the performance drop) this will commit will only drop while there is a request for the camera instead of having all of them updating every frame

Commit:
Viewport branchless assignation

Fixed gameViewport->bDisableWorldRendering assignation
Added comments explaining the reasons behind setting cameras to nodisplay

Commit:
Fix clang-format violation

rajat2004 commented:
Ah, I think I misunderstood the issue, I thought the FPS of `simGetImages` itself dropped, rather than the viewport. If possible, could you try the [image_benchmarker.py](https://github.com/microsoft/AirSim/blob/master/PythonClient/computer_vision/image_benchmarker.py) script with and without your fix, just to get a sense of performance improvement?

zimmy87 commented:
Thanks for submitting this fix @nznaza! Tested this locally using the repro steps for #1724 and #2967. Confirmed #1724 and #3504 (which is a duplicate of #1724). Unfortunately, I was still able to repro #2967 and I didn't see any significant performance change using the [image_benchmarker.py](https://github.com/microsoft/AirSim/blob/master/PythonClient/computer_vision/image_benchmarker.py) script, so I suspect #3070 still repros. Considering this does fix some of the issues, I'm gonna go ahead and merge this, we can address the other issues in a separate PR.
